Break ties in async scheduler

mrocklin opened the pull request:
This breaks ties when adding tasks to the scheduling stack in the async scheduler. 

This breaks ties by

1.  Using a Depth First Search to compute groups first.  Ties at this stage are broken by 2
2.  We select children based on the maximum over all children of the number of total data elements depending on that child

I'll think about how to clean up that definition in a bit and provide examples soon

cc @eriknw @jcrist @cowlicks this came out of our blackboard conversations

@ogrisel I believe that this will eventually cause the workflows we were running into to handle disk access much better.

mrocklin commented:
Playing with @ogrisel 's cross-validation problem has been really educational. 

The implementation in this PR seems to support the `reduction(axis=k)` case that @jcrist brought up quite well.  It also does a few artificial cases that I've added in tests, including, the fail case in the blogpost

![](http://mrocklin.github.com/blog/images/dask/fail-case.gif)

The depth first search approach is somewhat counter-productive with the cross-validation problem from @ogrisel .  Depth first search is hard to reason about in this case.

![mydask](https://cloud.githubusercontent.com/assets/306380/8583333/10b30d46-25cd-11e5-8738-13d2164ed391.png)

There are however other tiebreakers being used in the current system that *would* be a good fit, except they're currently being overwhelmed by `dfs`, which is really nice in array workloads.  This brings up the option of having pluggable tie-breaking schemes.  I'm not sure how much I like this.

I'll keep playing around for a while.

mrocklin commented:
Here is the current state of profile of the cross-validation workload.  It's much nicer than it used to be, but still not perfect.

https://rawgit.com/mrocklin/8ec0443c94da553fe00c/raw/ff7d8d0754d07f35086b08c0d21865a03b3edeac/profile.html

Warning, that's a 1.7MB page.  Might take a minute to load

mrocklin commented:
Looking at the resulting depth first search ordering it seems like we load about a quarter of the data at the front (this is bad) but the rest gets sprinkled through.

Note, this could be totally off because of how the dynamic system works, this is just what the static ordering dictates.

Here is a list of the dfs scores.  Keys with lower numbers will be run later if in a tie situation.  We want to see the high `from-cache` keys like `from-cache-99` receive very low scores (like 1).  This happens on all but a quarter of them which presumably resulted from some bad DFS ordering.

```python
In [7]: sorted(dfs.items(), key=lambda (k, v): v)
Out[7]: 
[(('cross_validate_fold_0-3', 'model', 75), 0),
 (('from-cache-5', 99), 1),
 (('cross_validate_fold_0-3', 'model', 74), 2),
 (('from-cache-5', 98), 3),
 (('cross_validate_fold_0-3', 'model', 73), 4),
 (('from-cache-5', 97), 5),
 (('cross_validate_fold_0-3', 'model', 72), 6),
 (('from-cache-5', 95), 7),
 (('cross_validate_fold_0-3', 'model', 71), 8),
 (('from-cache-5', 94), 9),
 (('cross_validate_fold_0-3', 'model', 70), 10),
 (('from-cache-5', 93), 11),
 (('cross_validate_fold_0-3', 'model', 69), 12),
 (('from-cache-5', 91), 13),
 (('cross_validate_fold_0-3', 'model', 68), 14),
 (('from-cache-5', 90), 15),
 (('cross_validate_fold_0-3', 'model', 67), 16),
 (('from-cache-5', 89), 17),
 (('cross_validate_fold_0-3', 'model', 66), 18),
 (('from-cache-5', 87), 19),
 (('cross_validate_fold_0-3', 'model', 65), 20),
 (('from-cache-5', 86), 21),
 (('cross_validate_fold_0-3', 'model', 64), 22),
 (('from-cache-5', 85), 23),
 (('cross_validate_fold_0-3', 'model', 63), 24),
 (('from-cache-5', 83), 25),
 (('cross_validate_fold_0-3', 'model', 62), 26),
 (('from-cache-5', 82), 27),
 (('cross_validate_fold_0-3', 'model', 61), 28),
 (('from-cache-5', 81), 29),
 (('cross_validate_fold_0-3', 'model', 60), 30),
 (('from-cache-5', 79), 31),
 (('cross_validate_fold_0-3', 'model', 59), 32),
 (('from-cache-5', 78), 33),
 (('cross_validate_fold_0-3', 'model', 58), 34),
 (('from-cache-5', 77), 35),
 (('cross_validate_fold_0-3', 'model', 57), 36),
 (('from-cache-5', 75), 37),
 (('cross_validate_fold_0-3', 'model', 56), 38),
 (('from-cache-5', 74), 39),
 (('cross_validate_fold_0-3', 'model', 55), 40),
 (('from-cache-5', 73), 41),
 (('cross_validate_fold_0-3', 'model', 54), 42),
 (('from-cache-5', 71), 43),
 (('cross_validate_fold_0-3', 'model', 53), 44),
 (('from-cache-5', 70), 45),
 (('cross_validate_fold_0-3', 'model', 52), 46),
 (('from-cache-5', 69), 47),
 (('cross_validate_fold_0-3', 'model', 51), 48),
 (('from-cache-5', 67), 49),
 (('cross_validate_fold_0-3', 'model', 50), 50),
 (('from-cache-5', 66), 51),
 (('cross_validate_fold_0-3', 'model', 49), 52),
 (('from-cache-5', 65), 53),
 (('cross_validate_fold_0-3', 'model', 48), 54),
 (('from-cache-5', 63), 55),
 (('cross_validate_fold_0-3', 'model', 47), 56),
 (('from-cache-5', 62), 57),
 (('cross_validate_fold_0-3', 'model', 46), 58),
 (('from-cache-5', 61), 59),
 (('cross_validate_fold_0-3', 'model', 45), 60),
 (('from-cache-5', 59), 61),
 (('cross_validate_fold_0-3', 'model', 44), 62),
 (('from-cache-5', 58), 63),
 (('cross_validate_fold_0-3', 'model', 43), 64),
 (('from-cache-5', 57), 65),
 (('cross_validate_fold_0-3', 'model', 42), 66),
 (('from-cache-5', 55), 67),
 (('cross_validate_fold_0-3', 'model', 41), 68),
 (('from-cache-5', 54), 69),
 (('cross_validate_fold_0-3', 'model', 40), 70),
 (('from-cache-5', 53), 71),
 (('cross_validate_fold_0-3', 'model', 39), 72),
 (('from-cache-5', 51), 73),
 (('cross_validate_fold_0-3', 'model', 38), 74),
 (('from-cache-5', 50), 75),
 (('cross_validate_fold_0-3', 'model', 37), 76),
 (('from-cache-5', 49), 77),
 (('cross_validate_fold_0-3', 'model', 36), 78),
 (('from-cache-5', 47), 79),
 (('cross_validate_fold_0-3', 'model', 35), 80),
 (('from-cache-5', 46), 81),
 (('cross_validate_fold_0-3', 'model', 34), 82),
 (('from-cache-5', 45), 83),
 (('cross_validate_fold_0-3', 'model', 33), 84),
 (('from-cache-5', 43), 85),
 (('cross_validate_fold_0-3', 'model', 32), 86),
 (('from-cache-5', 42), 87),
 (('cross_validate_fold_0-3', 'model', 31), 88),
 (('from-cache-5', 41), 89),
 (('cross_validate_fold_0-3', 'model', 30), 90),
 (('from-cache-5', 39), 91),
 (('cross_validate_fold_0-3', 'model', 29), 92),
 (('from-cache-5', 38), 93),
 (('cross_validate_fold_0-3', 'model', 28), 94),
 (('from-cache-5', 37), 95),
 (('cross_validate_fold_0-3', 'model', 27), 96),
 (('from-cache-5', 35), 97),
 (('cross_validate_fold_0-3', 'model', 26), 98),
 (('from-cache-5', 34), 99),
 (('cross_validate_fold_0-3', 'model', 25), 100),
 (('from-cache-5', 33), 101),
 (('cross_validate_fold_0-3', 'model', 24), 102),
 (('from-cache-5', 31), 103),
 (('cross_validate_fold_0-3', 'model', 23), 104),
 (('from-cache-5', 30), 105),
 (('cross_validate_fold_0-3', 'model', 22), 106),
 (('from-cache-5', 29), 107),
 (('cross_validate_fold_0-3', 'model', 21), 108),
 (('from-cache-5', 27), 109),
 (('cross_validate_fold_0-3', 'model', 20), 110),
 (('from-cache-5', 26), 111),
 (('cross_validate_fold_0-3', 'model', 19), 112),
 (('from-cache-5', 25), 113),
 (('cross_validate_fold_0-3', 'model', 18), 114),
 (('from-cache-5', 23), 115),
 (('cross_validate_fold_0-3', 'model', 17), 116),
 (('from-cache-5', 22), 117),
 (('cross_validate_fold_0-3', 'model', 16), 118),
 (('from-cache-5', 21), 119),
 (('cross_validate_fold_0-3', 'model', 15), 120),
 (('from-cache-5', 19), 121),
 (('cross_validate_fold_0-3', 'model', 14), 122),
 (('from-cache-5', 18), 123),
 (('cross_validate_fold_0-3', 'model', 13), 124),
 (('from-cache-5', 17), 125),
 (('cross_validate_fold_0-3', 'model', 12), 126),
 (('from-cache-5', 15), 127),
 (('cross_validate_fold_0-3', 'model', 11), 128),
 (('from-cache-5', 14), 129),
 (('cross_validate_fold_0-3', 'model', 10), 130),
 (('from-cache-5', 13), 131),
 (('cross_validate_fold_0-3', 'model', 9), 132),
 (('from-cache-5', 11), 133),
 (('cross_validate_fold_0-3', 'model', 8), 134),
 (('from-cache-5', 10), 135),
 (('cross_validate_fold_0-3', 'model', 7), 136),
 (('from-cache-5', 9), 137),
 (('cross_validate_fold_0-3', 'model', 6), 138),
 (('from-cache-5', 7), 139),
 (('cross_validate_fold_0-3', 'model', 5), 140),
 (('from-cache-5', 6), 141),
 (('cross_validate_fold_0-3', 'model', 4), 142),
 (('from-cache-5', 5), 143),
 (('cross_validate_fold_0-3', 'model', 3), 144),
 (('from-cache-5', 3), 145),
 (('cross_validate_fold_0-3', 'model', 2), 146),
 (('from-cache-5', 2), 147),
 (('cross_validate_fold_0-3', 'model', 1), 148),
 (('cross_validate_fold_0-3', 'model', 0), 149),
 (('from-cache-5', 1), 150),
 (('cross_validate_fold_0-3', 'validation', 23), 151),
 (('from-cache-5', 96), 152),
 (('cross_validate_fold_3-3', 'validation', 24), 153),
 (('cross_validate_fold_3-3', 'model', 75), 154),
 (('cross_validate_fold_3-3', 'model', 74), 155),
 (('cross_validate_fold_3-3', 'model', 73), 156),
 (('cross_validate_fold_3-3', 'model', 72), 157),
 (('cross_validate_fold_3-3', 'model', 71), 158),
 (('cross_validate_fold_3-3', 'model', 70), 159),
 (('from-cache-5', 92), 160),
 (('cross_validate_fold_3-3', 'model', 69), 161),
 (('cross_validate_fold_3-3', 'model', 68), 162),
 (('cross_validate_fold_3-3', 'model', 67), 163),
 (('from-cache-5', 88), 164),
 (('cross_validate_fold_3-3', 'model', 66), 165),
 (('cross_validate_fold_3-3', 'model', 65), 166),
 (('cross_validate_fold_3-3', 'model', 64), 167),
 (('from-cache-5', 84), 168),
 (('cross_validate_fold_3-3', 'model', 63), 169),
 (('cross_validate_fold_3-3', 'model', 62), 170),
 (('cross_validate_fold_3-3', 'model', 61), 171),
 (('from-cache-5', 80), 172),
 (('cross_validate_fold_3-3', 'model', 60), 173),
 (('cross_validate_fold_3-3', 'model', 59), 174),
 (('cross_validate_fold_3-3', 'model', 58), 175),
 (('from-cache-5', 76), 176),
 (('cross_validate_fold_3-3', 'model', 57), 177),
 (('cross_validate_fold_3-3', 'model', 56), 178),
 (('cross_validate_fold_3-3', 'model', 55), 179),
 (('from-cache-5', 72), 180),
 (('cross_validate_fold_3-3', 'model', 54), 181),
 (('cross_validate_fold_3-3', 'model', 53), 182),
 (('cross_validate_fold_3-3', 'model', 52), 183),
 (('from-cache-5', 68), 184),
 (('cross_validate_fold_3-3', 'model', 51), 185),
 (('cross_validate_fold_3-3', 'model', 50), 186),
 (('cross_validate_fold_3-3', 'model', 49), 187),
 (('from-cache-5', 64), 188),
 (('cross_validate_fold_3-3', 'model', 48), 189),
 (('cross_validate_fold_3-3', 'model', 47), 190),
 (('cross_validate_fold_3-3', 'model', 46), 191),
 (('from-cache-5', 60), 192),
 (('cross_validate_fold_3-3', 'model', 45), 193),
 (('cross_validate_fold_3-3', 'model', 44), 194),
 (('cross_validate_fold_3-3', 'model', 43), 195),
 (('from-cache-5', 56), 196),
 (('cross_validate_fold_3-3', 'model', 42), 197),
 (('cross_validate_fold_3-3', 'model', 41), 198),
 (('cross_validate_fold_3-3', 'model', 40), 199),
 (('from-cache-5', 52), 200),
 (('cross_validate_fold_3-3', 'model', 39), 201),
 (('cross_validate_fold_3-3', 'model', 38), 202),
 (('cross_validate_fold_3-3', 'model', 37), 203),
 (('from-cache-5', 48), 204),
 (('cross_validate_fold_3-3', 'model', 36), 205),
 (('cross_validate_fold_3-3', 'model', 35), 206),
 (('cross_validate_fold_3-3', 'model', 34), 207),
 (('from-cache-5', 44), 208),
 (('cross_validate_fold_3-3', 'model', 33), 209),
 (('cross_validate_fold_3-3', 'model', 32), 210),
 (('cross_validate_fold_3-3', 'model', 31), 211),
 (('from-cache-5', 40), 212),
 (('cross_validate_fold_3-3', 'model', 30), 213),
 (('cross_validate_fold_3-3', 'model', 29), 214),
 (('cross_validate_fold_3-3', 'model', 28), 215),
 (('from-cache-5', 36), 216),
 (('cross_validate_fold_3-3', 'model', 27), 217),
 (('cross_validate_fold_3-3', 'model', 26), 218),
 (('cross_validate_fold_3-3', 'model', 25), 219),
 (('from-cache-5', 32), 220),
 (('cross_validate_fold_3-3', 'model', 24), 221),
 (('cross_validate_fold_3-3', 'model', 23), 222),
 (('cross_validate_fold_3-3', 'model', 22), 223),
 (('from-cache-5', 28), 224),
 (('cross_validate_fold_3-3', 'model', 21), 225),
 (('cross_validate_fold_3-3', 'model', 20), 226),
 (('cross_validate_fold_3-3', 'model', 19), 227),
 (('from-cache-5', 24), 228),
 (('cross_validate_fold_3-3', 'model', 18), 229),
 (('cross_validate_fold_3-3', 'model', 17), 230),
 (('cross_validate_fold_3-3', 'model', 16), 231),
 (('from-cache-5', 20), 232),
 (('cross_validate_fold_3-3', 'model', 15), 233),
 (('cross_validate_fold_3-3', 'model', 14), 234),
 (('cross_validate_fold_3-3', 'model', 13), 235),
 (('from-cache-5', 16), 236),
 (('cross_validate_fold_3-3', 'model', 12), 237),
 (('cross_validate_fold_3-3', 'model', 11), 238),
 (('cross_validate_fold_3-3', 'model', 10), 239),
 (('from-cache-5', 12), 240),
 (('cross_validate_fold_3-3', 'model', 9), 241),
 (('cross_validate_fold_3-3', 'model', 8), 242),
 (('cross_validate_fold_3-3', 'model', 7), 243),
 (('from-cache-5', 8), 244),
 (('cross_validate_fold_3-3', 'model', 6), 245),
 (('cross_validate_fold_3-3', 'model', 5), 246),
 (('cross_validate_fold_3-3', 'model', 4), 247),
 (('from-cache-5', 4), 248),
 (('cross_validate_fold_3-3', 'model', 3), 249),
 (('cross_validate_fold_3-3', 'model', 2), 250),
 (('cross_validate_fold_3-3', 'model', 1), 251),
 (('cross_validate_fold_3-3', 'model', 0), 252),
 (('from-cache-5', 0), 253),
 (('cross_validate_fold_2-3', 'validation', 24), 254),
 (('cross_validate_fold_2-3', 'model', 74), 255),
 (('cross_validate_fold_2-3', 'model', 73), 256),
 (('cross_validate_fold_2-3', 'model', 72), 257),
 (('cross_validate_fold_2-3', 'model', 71), 258),
 (('cross_validate_fold_2-3', 'model', 70), 259),
 (('cross_validate_fold_2-3', 'model', 69), 260),
 (('cross_validate_fold_2-3', 'model', 68), 261),
 (('cross_validate_fold_2-3', 'model', 67), 262),
 (('cross_validate_fold_2-3', 'model', 66), 263),
 (('cross_validate_fold_2-3', 'model', 65), 264),
 (('cross_validate_fold_2-3', 'model', 64), 265),
 (('cross_validate_fold_2-3', 'model', 63), 266),
 (('cross_validate_fold_2-3', 'model', 62), 267),
 (('cross_validate_fold_2-3', 'model', 61), 268),
 (('cross_validate_fold_2-3', 'model', 60), 269),
 (('cross_validate_fold_2-3', 'model', 59), 270),
 (('cross_validate_fold_2-3', 'model', 58), 271),
 (('cross_validate_fold_2-3', 'model', 57), 272),
 (('cross_validate_fold_2-3', 'model', 56), 273),
 (('cross_validate_fold_2-3', 'model', 55), 274),
 (('cross_validate_fold_2-3', 'model', 54), 275),
 (('cross_validate_fold_2-3', 'model', 53), 276),
 (('cross_validate_fold_2-3', 'model', 52), 277),
 (('cross_validate_fold_2-3', 'model', 51), 278),
 (('cross_validate_fold_2-3', 'model', 50), 279),
 (('cross_validate_fold_2-3', 'model', 49), 280),
 (('cross_validate_fold_2-3', 'model', 48), 281),
 (('cross_validate_fold_2-3', 'model', 47), 282),
 (('cross_validate_fold_2-3', 'model', 46), 283),
 (('cross_validate_fold_2-3', 'model', 45), 284),
 (('cross_validate_fold_2-3', 'model', 44), 285),
 (('cross_validate_fold_2-3', 'model', 43), 286),
 (('cross_validate_fold_2-3', 'model', 42), 287),
 (('cross_validate_fold_2-3', 'model', 41), 288),
 (('cross_validate_fold_2-3', 'model', 40), 289),
 (('cross_validate_fold_2-3', 'model', 39), 290),
 (('cross_validate_fold_2-3', 'model', 38), 291),
 (('cross_validate_fold_2-3', 'model', 37), 292),
 (('cross_validate_fold_2-3', 'model', 36), 293),
 (('cross_validate_fold_2-3', 'model', 35), 294),
 (('cross_validate_fold_2-3', 'model', 34), 295),
 (('cross_validate_fold_2-3', 'model', 33), 296),
 (('cross_validate_fold_2-3', 'model', 32), 297),
 (('cross_validate_fold_2-3', 'model', 31), 298),
 (('cross_validate_fold_2-3', 'model', 30), 299),
 (('cross_validate_fold_2-3', 'model', 29), 300),
 (('cross_validate_fold_2-3', 'model', 28), 301),
 (('cross_validate_fold_2-3', 'model', 27), 302),
 (('cross_validate_fold_2-3', 'model', 26), 303),
 (('cross_validate_fold_2-3', 'model', 25), 304),
 (('cross_validate_fold_2-3', 'model', 24), 305),
 (('cross_validate_fold_2-3', 'model', 23), 306),
 (('cross_validate_fold_2-3', 'model', 22), 307),
 (('cross_validate_fold_2-3', 'model', 21), 308),
 (('cross_validate_fold_2-3', 'model', 20), 309),
 (('cross_validate_fold_2-3', 'model', 19), 310),
 (('cross_validate_fold_2-3', 'model', 18), 311),
 (('cross_validate_fold_2-3', 'model', 17), 312),
 (('cross_validate_fold_2-3', 'model', 16), 313),
 (('cross_validate_fold_2-3', 'model', 15), 314),
 (('cross_validate_fold_2-3', 'model', 14), 315),
 (('cross_validate_fold_2-3', 'model', 13), 316),
 (('cross_validate_fold_2-3', 'model', 12), 317),
 (('cross_validate_fold_2-3', 'model', 11), 318),
 (('cross_validate_fold_2-3', 'model', 10), 319),
 (('cross_validate_fold_2-3', 'model', 9), 320),
 (('cross_validate_fold_2-3', 'model', 8), 321),
 (('cross_validate_fold_2-3', 'model', 7), 322),
 (('cross_validate_fold_2-3', 'model', 6), 323),
 (('cross_validate_fold_2-3', 'model', 5), 324),
 (('cross_validate_fold_2-3', 'model', 4), 325),
 (('cross_validate_fold_2-3', 'model', 3), 326),
 (('cross_validate_fold_2-3', 'model', 2), 327),
 (('cross_validate_fold_2-3', 'model', 1), 328),
 (('cross_validate_fold_2-3', 'model', 0), 329),
 (('cross_validate_fold_1-3', 'model', 75), 330),
 (('cross_validate_fold_1-3', 'model', 74), 331),
 (('cross_validate_fold_1-3', 'model', 73), 332),
 (('cross_validate_fold_1-3', 'model', 72), 333),
 (('cross_validate_fold_1-3', 'model', 71), 334),
 (('cross_validate_fold_1-3', 'model', 70), 335),
 (('cross_validate_fold_1-3', 'model', 69), 336),
 (('cross_validate_fold_1-3', 'model', 68), 337),
 (('cross_validate_fold_1-3', 'model', 67), 338),
 (('cross_validate_fold_1-3', 'model', 66), 339),
 (('cross_validate_fold_1-3', 'model', 65), 340),
 (('cross_validate_fold_1-3', 'model', 64), 341),
 (('cross_validate_fold_1-3', 'model', 63), 342),
 (('cross_validate_fold_1-3', 'model', 62), 343),
 (('cross_validate_fold_1-3', 'model', 61), 344),
 (('cross_validate_fold_1-3', 'model', 60), 345),
 (('cross_validate_fold_1-3', 'model', 59), 346),
 (('cross_validate_fold_1-3', 'model', 58), 347),
 (('cross_validate_fold_1-3', 'model', 57), 348),
 (('cross_validate_fold_1-3', 'model', 56), 349),
 (('cross_validate_fold_1-3', 'model', 55), 350),
 (('cross_validate_fold_1-3', 'model', 54), 351),
 (('cross_validate_fold_1-3', 'model', 53), 352),
 (('cross_validate_fold_1-3', 'model', 52), 353),
 (('cross_validate_fold_1-3', 'model', 51), 354),
 (('cross_validate_fold_1-3', 'model', 50), 355),
 (('cross_validate_fold_1-3', 'model', 49), 356),
 (('cross_validate_fold_1-3', 'model', 48), 357),
 (('cross_validate_fold_1-3', 'model', 47), 358),
 (('cross_validate_fold_1-3', 'model', 46), 359),
 (('cross_validate_fold_1-3', 'model', 45), 360),
 (('cross_validate_fold_1-3', 'model', 44), 361),
 (('cross_validate_fold_1-3', 'model', 43), 362),
 (('cross_validate_fold_1-3', 'model', 42), 363),
 (('cross_validate_fold_1-3', 'model', 41), 364),
 (('cross_validate_fold_1-3', 'model', 40), 365),
 (('cross_validate_fold_1-3', 'model', 39), 366),
 (('cross_validate_fold_1-3', 'model', 38), 367),
 (('cross_validate_fold_1-3', 'model', 37), 368),
 (('cross_validate_fold_1-3', 'model', 36), 369),
 (('cross_validate_fold_1-3', 'model', 35), 370),
 (('cross_validate_fold_1-3', 'model', 34), 371),
 (('cross_validate_fold_1-3', 'model', 33), 372),
 (('cross_validate_fold_1-3', 'model', 32), 373),
 (('cross_validate_fold_1-3', 'model', 31), 374),
 (('cross_validate_fold_1-3', 'model', 30), 375),
 (('cross_validate_fold_1-3', 'model', 29), 376),
 (('cross_validate_fold_1-3', 'model', 28), 377),
 (('cross_validate_fold_1-3', 'model', 27), 378),
 (('cross_validate_fold_1-3', 'model', 26), 379),
 (('cross_validate_fold_1-3', 'model', 25), 380),
 (('cross_validate_fold_1-3', 'model', 24), 381),
 (('cross_validate_fold_1-3', 'model', 23), 382),
 (('cross_validate_fold_1-3', 'model', 22), 383),
 (('cross_validate_fold_1-3', 'model', 21), 384),
 (('cross_validate_fold_1-3', 'model', 20), 385),
 (('cross_validate_fold_1-3', 'model', 19), 386),
 (('cross_validate_fold_1-3', 'model', 18), 387),
 (('cross_validate_fold_1-3', 'model', 17), 388),
 (('cross_validate_fold_1-3', 'model', 16), 389),
 (('cross_validate_fold_1-3', 'model', 15), 390),
 (('cross_validate_fold_1-3', 'model', 14), 391),
 (('cross_validate_fold_1-3', 'model', 13), 392),
 (('cross_validate_fold_1-3', 'model', 12), 393),
 (('cross_validate_fold_1-3', 'model', 11), 394),
 (('cross_validate_fold_1-3', 'model', 10), 395),
 (('cross_validate_fold_1-3', 'model', 9), 396),
 (('cross_validate_fold_1-3', 'model', 8), 397),
 (('cross_validate_fold_1-3', 'model', 7), 398),
 (('cross_validate_fold_1-3', 'model', 6), 399),
 (('cross_validate_fold_1-3', 'model', 5), 400),
 (('cross_validate_fold_1-3', 'model', 4), 401),
 (('cross_validate_fold_1-3', 'model', 3), 402),
 (('cross_validate_fold_1-3', 'model', 2), 403),
 (('cross_validate_fold_1-3', 'model', 1), 404),
 (('cross_validate_fold_1-3', 'model', 0), 405),
 (('cross_validate_fold_2-3', 'model', 75), 406),
 (('cross_validate_fold_1-3', 'validation', 24), 407),
 (('cross_validate_fold_0-3', 'validation', 22), 408),
 (('cross_validate_fold_3-3', 'validation', 23), 409),
 (('cross_validate_fold_2-3', 'validation', 23), 410),
 (('cross_validate_fold_1-3', 'validation', 23), 411),
 (('cross_validate_fold_0-3', 'validation', 21), 412),
 (('cross_validate_fold_3-3', 'validation', 22), 413),
 (('cross_validate_fold_2-3', 'validation', 22), 414),
 (('cross_validate_fold_1-3', 'validation', 22), 415),
 (('cross_validate_fold_0-3', 'validation', 20), 416),
 (('cross_validate_fold_3-3', 'validation', 21), 417),
 (('cross_validate_fold_2-3', 'validation', 21), 418),
 (('cross_validate_fold_1-3', 'validation', 21), 419),
 (('cross_validate_fold_0-3', 'validation', 19), 420),
 (('cross_validate_fold_3-3', 'validation', 20), 421),
 (('cross_validate_fold_2-3', 'validation', 20), 422),
 (('cross_validate_fold_1-3', 'validation', 20), 423),
 (('cross_validate_fold_0-3', 'validation', 18), 424),
 (('cross_validate_fold_3-3', 'validation', 19), 425),
 (('cross_validate_fold_2-3', 'validation', 19), 426),
 (('cross_validate_fold_0-3', 'validation', 17), 427),
 (('cross_validate_fold_1-3', 'validation', 19), 428),
 (('cross_validate_fold_3-3', 'validation', 18), 429),
 (('cross_validate_fold_2-3', 'validation', 18), 430),
 (('cross_validate_fold_1-3', 'validation', 18), 431),
 (('cross_validate_fold_0-3', 'validation', 16), 432),
 (('cross_validate_fold_3-3', 'validation', 17), 433),
 (('cross_validate_fold_2-3', 'validation', 17), 434),
 (('cross_validate_fold_1-3', 'validation', 17), 435),
 (('cross_validate_fold_0-3', 'validation', 15), 436),
 (('cross_validate_fold_3-3', 'validation', 16), 437),
 (('cross_validate_fold_2-3', 'validation', 16), 438),
 (('cross_validate_fold_1-3', 'validation', 16), 439),
 (('cross_validate_fold_0-3', 'validation', 14), 440),
 (('cross_validate_fold_3-3', 'validation', 15), 441),
 (('cross_validate_fold_2-3', 'validation', 15), 442),
 (('cross_validate_fold_1-3', 'validation', 15), 443),
 (('cross_validate_fold_0-3', 'validation', 13), 444),
 (('cross_validate_fold_3-3', 'validation', 14), 445),
 (('cross_validate_fold_2-3', 'validation', 14), 446),
 (('cross_validate_fold_1-3', 'validation', 14), 447),
 (('cross_validate_fold_0-3', 'validation', 12), 448),
 (('cross_validate_fold_3-3', 'validation', 13), 449),
 (('cross_validate_fold_2-3', 'validation', 13), 450),
 (('cross_validate_fold_1-3', 'validation', 13), 451),
 (('cross_validate_fold_0-3', 'validation', 11), 452),
 (('cross_validate_fold_3-3', 'validation', 12), 453),
 (('cross_validate_fold_2-3', 'validation', 12), 454),
 (('cross_validate_fold_1-3', 'validation', 12), 455),
 (('cross_validate_fold_0-3', 'validation', 10), 456),
 (('cross_validate_fold_3-3', 'validation', 11), 457),
 (('cross_validate_fold_2-3', 'validation', 11), 458),
 (('cross_validate_fold_0-3', 'validation', 9), 459),
 (('cross_validate_fold_1-3', 'validation', 11), 460),
 (('cross_validate_fold_3-3', 'validation', 10), 461),
 (('cross_validate_fold_2-3', 'validation', 10), 462),
 (('cross_validate_fold_0-3', 'validation', 8), 463),
 (('cross_validate_fold_1-3', 'validation', 10), 464),
 (('cross_validate_fold_3-3', 'validation', 9), 465),
 (('cross_validate_fold_2-3', 'validation', 9), 466),
 (('cross_validate_fold_1-3', 'validation', 9), 467),
 (('cross_validate_fold_0-3', 'validation', 7), 468),
 (('cross_validate_fold_3-3', 'validation', 8), 469),
 (('cross_validate_fold_2-3', 'validation', 8), 470),
 (('cross_validate_fold_1-3', 'validation', 8), 471),
 (('cross_validate_fold_0-3', 'validation', 6), 472),
 (('cross_validate_fold_3-3', 'validation', 7), 473),
 (('cross_validate_fold_2-3', 'validation', 7), 474),
 (('cross_validate_fold_1-3', 'validation', 7), 475),
 (('cross_validate_fold_0-3', 'validation', 5), 476),
 (('cross_validate_fold_3-3', 'validation', 6), 477),
 (('cross_validate_fold_2-3', 'validation', 6), 478),
 (('cross_validate_fold_1-3', 'validation', 6), 479),
 (('cross_validate_fold_0-3', 'validation', 4), 480),
 (('cross_validate_fold_3-3', 'validation', 5), 481),
 (('cross_validate_fold_2-3', 'validation', 5), 482),
 (('cross_validate_fold_1-3', 'validation', 5), 483),
 (('cross_validate_fold_0-3', 'validation', 3), 484),
 (('cross_validate_fold_3-3', 'validation', 4), 485),
 (('cross_validate_fold_2-3', 'validation', 4), 486),
 (('cross_validate_fold_1-3', 'validation', 4), 487),
 (('cross_validate_fold_0-3', 'validation', 2), 488),
 (('cross_validate_fold_3-3', 'validation', 3), 489),
 (('cross_validate_fold_2-3', 'validation', 3), 490),
 (('cross_validate_fold_0-3', 'validation', 1), 491),
 (('cross_validate_fold_1-3', 'validation', 3), 492),
 (('cross_validate_fold_3-3', 'validation', 2), 493),
 (('cross_validate_fold_2-3', 'validation', 2), 494),
 (('cross_validate_fold_0-3', 'validation', 0), 495),
 (('cross_validate_fold_1-3', 'validation', 2), 496),
 (('cross_validate_fold_3-3', 'validation', 1), 497),
 (('cross_validate_fold_2-3', 'validation', 1), 498),
 (('cross_validate_fold_1-3', 'validation', 1), 499),
 (('cross_validate_fold_3-3', 'validation', 0), 500),
 (('cross_validate_fold_2-3', 'validation', 0), 501),
 (('cross_validate_fold_1-3', 'validation', 0), 502)]

```

mrocklin commented:
Still failing hard on the array case.

eriknw commented:
I see.  Interesting graph, thanks for sharing.

I still like having a static scheduler that returns a natural ordering of keys to calculate (such as `{'a': 3, 'c': 2, 'b': 1, 'd': 0}` as in the docstring example). This is intuitive, as simple as possible an output to create, and follows in the tradition of list scheduling techniques.  This can then feed into one of possibly many dynamic schedulers, which can be developed entirely independently.

The current approach of the dynamic scheduler is to greedily choose tasks that free memory, which is generally a good idea.  Another useful behavior when there are multiple workers may be to limit premature task execution, which are those tasks whose results will be guaranteed to sit in memory until the primary dfs chain catches up.  Instead, the scheduler should "step wide" instead of "stepping deep" and begin running the next tall dfs branch (if available; if not, wait).  Heh, to make this efficient, I envision something similar to "Chutes and Ladders" combined with simple metrics.

Of course, for the case of cross-validation with a dfs static scheduler, the dynamic scheduler would need to disregard and override the static scheduling information to have the desired behavior.  This suggests the static scheduler should be better, which you alluded to.  I don't think it's a terrible idea to allow and support advanced users to construct their own scheduling policies.  As we continue to explore this space, it may make sense to create `dask.scheduling.metrics` (or find and use--or make--a library that does this).  Aargh, if only it weren't so late already... :)

ogrisel commented:
I have been thinking a bit about scheduling issues while commuting this morning. Here is a brain dump:

- leaf nodes of the original dag are all data loading operations [1]: they should preferably not be scheduled as long as at least one worker is about to starve otherwise you run the risk of overloading the memory usage of the system for no benefit,
- you still need to schedule some leafs so as to start. But instead of scheduling leafs arbitrarily, one should schedule leafs in an order that allows the intermediate nodes to be consumed first,
- breaking ties by scheduling deepest leaf first seems reasonable to me: it should make it possible to minimize the number of intermediate values to store (or spill to disk) while waiting for remaining computation to move up,
- maybe for scheduling it would help considering the sub-dag of processing nodes first (that is the dag that remains when you strip all the data loading leafs from the original dag) and make scheduling decisions based on that dag only and then complete that initial schedule skeleton by inserting any leaf node of the original dag whenever they are required as a dependency of a process node.

[1] At least this is the case in data processing programs which I think is the primary use case for the dask projects. For simulation operations leaf nodes could be small input parameters that do no require allocation of large in memory buffers them-selves.

mrocklin commented:
> Instead, the scheduler should "step wide" instead of "stepping deep" and begin running the next tall dfs branch (if available; if not, wait)

@eriknw can you give an example of this?  I'm not certain what you mean here.

> I don't think it's a terrible idea to allow and support advanced users to construct their own scheduling policies

I agree that that doesn't sound terrible.  It sounds kind of great actually.

mrocklin commented:
> leaf nodes of the original dag are all data loading operations [1]: they should preferably not be scheduled as long as at least one worker is about to starve otherwise you run the risk of overloading the memory usage of the system for no benefit,

@ogrisel this is current behavior.  The dynamic scheduler maintains a LIFO stack of ready-to-run tasks.  Leaf nodes are added at the beginning of the computation and so, once they make intermediate tasks available, other leaf nodes always wait.  By doing this LIFO trick we extend this heuristic beyond leaf nodes to other similar situations.

> you still need to schedule some leafs so as to start. But instead of scheduling leafs arbitrarily, one should schedule leafs in an order that allows the intermediate process nodes to be executed first,

This is the point of this pull request which effectively orders the leaf nodes (and all other nodes) by the amount of data that depends on them both in regards to depth and width.  Hopefully this removes bottleneck tasks quickly and encourages vertical flow in the future.

> [1] At least this is the case in data processing programs which I think is the primary use case for the dask projects. For simulation operations leaf nodes could be small input parameters that do no require allocation of large in memory buffers them-selves.

I agree that selecting tasks to minimize memory usage is probably optimal for the set of tasks that we have considered so far.

mrocklin commented:
I'm tempted to merge this somewhat soon and iterate in another PR.  I think that the general approach of ordering the graph ahead of time and using that order to break ties is a decent one.  We may still want to play with the exact ordering function.

### TODO

- [ ] Verify that this does not introduce significant overhead
- [ ] Check in with @jcrist that he is ok with this approach.

mrocklin commented:
Calling order on dask graphs seems pretty cheap

### order on ogrisel's crossval dsk is cheap
```python
In [4]: %time _ = order(dsk)
CPU times: user 20 ms, sys: 0 ns, total: 20 ms
Wall time: 16.8 ms

In [5]: len(dsk)
Out[5]: 503
```

### order on common case dask array code isn't bad 

```python
In [1]: import dask.array as da

In [2]: x = da.ones((1000, 1000), chunks=(5, 5))

In [3]: %prun -s cumtime x.sum().compute()
```
```
         27332943 function calls (26451934 primitive calls) in 40.394 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.014    0.014   40.394   40.394 <string>:1(<module>)
        1    0.004    0.004   39.173   39.173 core.py:708(compute)
        1    0.006    0.006   39.170   39.170 core.py:496(compute)
        1    0.012    0.012   39.156   39.156 core.py:1223(get)
        1    0.014    0.014   26.203   26.203 threaded.py:22(get)
        1    0.674    0.674   26.190   26.190 async.py:380(get_async)
   224989   13.713    0.000   13.713    0.000 {method 'acquire' of 'thread.lock' objects}
        1    0.022    0.022   12.940   12.940 optimization.py:11(optimize)
    40001    0.663    0.000   12.282    0.000 Queue.py:150(get)
    40001    0.698    0.000    9.355    0.000 async.py:442(fire_task)
    16246    0.473    0.000    8.736    0.001 threading.py:309(wait)
   560009    4.023    0.000    7.709    0.000 core.py:177(get_dependencies)
    40001    0.288    0.000    7.039    0.000 pool.py:289(apply_async)
        1    0.048    0.048    5.400    5.400 optimization.py:45(remove_full_slices)
    40001    0.474    0.000    5.382    0.000 Queue.py:107(put)
        1    0.097    0.097    4.941    4.941 optimize.py:218(dealias)
  3722831    2.290    0.000    2.976    0.000 core.py:27(istask)
       15    0.067    0.004    2.874    0.192 {method 'update' of 'dict' objects}
        7    0.000    0.000    2.848    0.407 dicttoolz.py:72(valmap)
    40001    0.025    0.000    2.807    0.000 rewrite.py:312(rewrite)
160202/40001    0.395    0.000    2.782    0.000 rewrite.py:360(_bottom_up)
    80002    0.937    0.000    2.506    0.000 threading.py:373(notify)
       42    2.296    0.055    2.296    0.055 {method 'items' of 'dict' objects}
160002/80002    0.063    0.000    2.065    0.000 rewrite.py:362(<genexpr>)
   160202    0.121    0.000    1.996    0.000 rewrite.py:301(_rewrite)
        1    0.062    0.062    1.947    1.947 optimize.py:161(inline_functions)
   160202    0.307    0.000    1.875    0.000 rewrite.py:280(iter_matches)
        5    0.475    0.095    1.654    0.331 core.py:248(reverse_dict)
        1    0.421    0.421    1.541    1.541 optimize.py:36(fuse)
    96248    0.354    0.000    1.521    0.000 threading.py:300(_is_owned)
        1    0.002    0.002    1.445    1.445 async.py:694(order)
   160202    0.720    0.000    1.387    0.000 rewrite.py:372(_match)
    40001    0.287    0.000    1.368    0.000 pool.py:528(__init__)
   920000    0.647    0.000    1.344    0.000 core.py:149(_deps)
        1    0.066    0.066    1.318    1.318 async.py:135(start_state_from_dask)
  5405889    1.070    0.000    1.303    0.000 {isinstance}
    80002    0.074    0.000    1.277    0.000 optimize.py:251(<genexpr>)
        1    0.000    0.000    1.206    1.206 core.py:942(sum)
        1    0.003    0.003    1.206    1.206 reductions.py:54(sum)
        1    0.002    0.002    1.202    1.202 reductions.py:16(reduction)
        2    0.008    0.004    1.201    0.600 core.py:1134(atop)
        1    0.113    0.113    1.182    1.182 optimize.py:9(cull)
        2    0.214    0.107    1.174    0.587 core.py:231(top)
        1    0.135    0.135    1.148    1.148 optimize.py:110(inline)
360202/40001    0.483    0.000    1.065    0.000 optimize.py:192(functions_of)
    40004    0.193    0.000    1.054    0.000 threading.py:242(Condition)
    80002    0.066    0.000    1.043    0.000 optimize.py:242(<genexpr>)
    40001    0.450    0.000    0.951    0.000 async.py:298(finish_task)
160203/40002    0.153    0.000    0.884    0.000 {map}
        1    0.016    0.016    0.876    0.876 async.py:577(get_deps)
    40004    0.796    0.000    0.861    0.000 threading.py:260(__init__)
    16246    0.050    0.000    0.845    0.000 threading.py:297(_acquire_restore)
    40001    0.024    0.000    0.758    0.000 core.py:158(zero_broadcast_dimensions)
80201/40001    0.094    0.000    0.734    0.000 utils.py:22(deepmap)
```

Adding ordering isn't free, but it's well beneath other costs.  about half of its cost is through `get_depdendencies`, which we could/should really do once per computation but end up doing several times.

jcrist commented on dask/async.py:
```diff
@@ -567,3 +572,139 @@ def sortkey(item):
+def child_max(dependencies, dependents, scores):
+    """
+
+    The maximum of all child scores minus a half.
```
Extra line

mrocklin replied:
Fixed

jcrist commented on dask/async.py:
```diff
@@ -567,3 +572,139 @@ def sortkey(item):
+def dfs(dependencies, dependents, key=lambda x: x):
+    """
+
+    >>> dsk = {'a': 1, 'b': 2, 'c': (inc, 'a'), 'd': (add, 'b', 'c')}
```
Extra line??

mrocklin replied:
This is a sign that I wanted a doctest but couldn't yet think of a title line for the docstring itself.  More a peculiarity of my doctest-heavy programming process.

I've added a title line and general prose both here and above.

jcrist commented:
Overall, looks good to me, we can tune the strategy later. I might move the ordering functions to a separate file (async is getting a little long, and they may be useful elsewhere), but I don't feel strongly about that. Feel free to merge.

mrocklin commented:
OK, I've moved the ordering code to `order.py` and added narrative documentation within the file.  Hopefully it's clear enough for others to pick up if it needs to be tweaked in the future.
